pkg/planner/util/null_misc.go

```diff
@@ -18,76 +18,98 @@ import (
-	return specialNullRejectedCase1(ctx, schema, expr) // only 1 case now
+// isNullRejectedInList checks null filter for IN list using OR logic.
+// Reason is that null filtering through evaluation by isNullRejectedSimpleExpr
+// has problems with IN list.
+func isNullRejectedInList(ctx base.PlanContext, expr *expression.ScalarFunction, innerSchema *expression.Schema) bool {
+	for i, arg := range expr.GetArgs() {
+		if i > 0 {
+			newArgs := make([]expression.Expression, 0, 2)
+			newArgs = append(newArgs, expr.GetArgs()[0])
+			newArgs = append(newArgs, arg)
+			eQCondition, err := expression.NewFunction(ctx.GetExprCtx(), ast.EQ, expr.GetType(), newArgs...)
+			if err != nil {
+				return false
+			}
+			if !(isNullRejectedSimpleExpr(ctx, innerSchema, eQCondition)) {
+				return false
+			}
+		}
+	}
+	return true
 }
 
-// specialNullRejectedCase1 is mainly for #49616.
-// Case1 specially handles `null-rejected OR (null-rejected AND {others})`, then no matter what the result
-// of `{others}` is (True, False or Null), the result of this predicate is null, so this predicate is null-rejected.
-func specialNullRejectedCase1(ctx context.PlanContext, schema *expression.Schema, expr expression.Expression) bool {
-	isFunc := func(e expression.Expression, lowerFuncName string) *expression.ScalarFunction {
-		f, ok := e.(*expression.ScalarFunction)
-		if !ok {
-			return nil
-		}
-		if f.FuncName.L == lowerFuncName {
-			return f
+// IsNullRejected takes care of complex predicates like this:
+// IsNullRejected(A OR B) = IsNullRejected(A) AND IsNullRejected(B)
+// IsNullRejected(A AND B) = IsNullRejected(A) OR IsNullRejected(B)
+func IsNullRejected(ctx base.PlanContext, innerSchema *expression.Schema, predicate expression.Expression) bool {
+	predicate = expression.PushDownNot(ctx.GetNullRejectCheckExprCtx(), predicate)
+	if expression.ContainOuterNot(predicate) {
+		return false
+	}
+
+	switch expr := predicate.(type) {
+	case *expression.ScalarFunction:
+		if expr.FuncName.L == ast.LogicAnd {
+			if IsNullRejected(ctx, innerSchema, expr.GetArgs()[0]) {
+				return true
+			}
+			return IsNullRejected(ctx, innerSchema, expr.GetArgs()[1])
+		} else if expr.FuncName.L == ast.LogicOr {
+			if !(IsNullRejected(ctx, innerSchema, expr.GetArgs()[0])) {
+				return false
+			}
+			return IsNullRejected(ctx, innerSchema, expr.GetArgs()[1])
+		} else if expr.FuncName.L == ast.In {
+			return isNullRejectedInList(ctx, expr, innerSchema)
+		} else {
+			return isNullRejectedSimpleExpr(ctx, innerSchema, expr)
 		}
-		return nil
+	default:
+		return isNullRejectedSimpleExpr(ctx, innerSchema, predicate)
 	}
-	orFunc := isFunc(expr, ast.LogicOr)
-	if orFunc == nil {
+}
+
+// isNullRejectedSimpleExpr check whether a condition is null-rejected
+// A condition would be null-rejected in one of following cases:
+// If it is a predicate containing a reference to an inner table that evaluates to UNKNOWN or FALSE
```

Comment:
i think the comment should be refined as well, "inner table" has no ideas with

Reply:
added explanation to inner table as "null producing side"